File: Rust/rust_aoc/src/year_2015/d22_boss_battle_2.rs
use std::fmt;

pub fn part_one(lines: &Vec<String>) -> String {
    let boss = Boss::parse(lines);
    let player = Player{hp: 50, mana: 500, armor: 0};
    let initial_state = State{player: player, boss: boss, cast_spells: Vec::new(), mana_spent: 0};

    find_best_mana(initial_state, false).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let boss = Boss::parse(lines);
    let player = Player{hp: 50, mana: 500, armor: 0};
    let initial_state = State{player: player, boss: boss, cast_spells: Vec::new(), mana_spent: 0};

    find_best_mana(initial_state, true).to_string()
}

#[derive(Clone, Copy, Eq)]
struct Spell<'a> {
    name: &'a str,
    cost: i32,
    damage: i32,
    heal: i32,
    armor: i32,
    restore_mana: i32,
    duration: i32
}

impl<'a> std::cmp::PartialEq for Spell<'a> {
    fn eq(&self, other: &Spell) -> bool {
       self.name == other.name
    }
}
impl<'a> fmt::Debug for Spell<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl<'a> Spell<'a> {
    fn new() -> Self {
        Spell { name: "", cost: 0, damage: 0, heal: 0, armor: 0, restore_mana: 0, duration: 0}
    }
}

#[derive(Clone)]
struct Player { hp: i32, mana: i32, armor: i32 }
#[derive(Clone)]
struct Boss { hp: i32, atk: i32 }

impl Boss { 
    fn parse(lines: &Vec<String>) -> Self { 
        let hp = lines[0].split(" ").last().unwrap().parse().unwrap();
        let atk = lines[1].split(" ").last().unwrap().parse().unwrap();
        Self{hp, atk} 
    } 
}
#[derive(Clone)]
struct State {
    player: Player,
    boss: Boss,
    cast_spells: Vec<Spell<'static>>,
    mana_spent: i32
}

const ALL_SPELLS: [&'static Spell; 5] = [
    &Spell{name: "MagicMissle", cost: 53, damage: 4, heal: 0, armor: 0, restore_mana: 0, duration: 1},
    &Spell{name: "Drain", cost: 73, damage: 2, heal: 2, armor: 0, restore_mana: 0, duration: 1},
    &Spell{name: "Shield", cost: 113, damage: 0, heal: 0, armor: 7, restore_mana: 0, duration: 6},
    &Spell{name: "Poison", cost: 173, damage: 3, heal: 0, armor: 0, restore_mana: 0, duration: 6},
    &Spell{name: "Recharge", cost: 229, damage: 0, heal: 0, armor: 0, restore_mana: 101, duration: 5}
    ];

fn tick_spells(state: &mut State) {
    let total = state.cast_spells.iter().fold(Spell::new(), |mut res, e| {
        res.damage += e.damage;
        res.heal += e.heal;
        res.armor += e.armor;
        res.restore_mana += e.restore_mana;
        res
    });
    state.player.hp += total.heal;
    state.player.armor = total.armor;
    state.player.mana += total.restore_mana;
    state.boss.hp -= total.damage;

    state.cast_spells = state.cast_spells.iter().map(|s| Spell{duration: s.duration - 1, ..*s}).filter(|s| s.duration > 0).collect();
}

fn check_match_over(state: &State, best: &mut i32) -> bool {
    if state.boss.hp <= 0 {
        if state.mana_spent < *best {
            *best = state.mana_spent;
        }
        return true;
    }
    if state.player.hp <= 0 {
        return true;
    }
    false
}

fn get_next_spells (state: State) -> Vec<Spell<'static>> {
    ALL_SPELLS.iter().filter(|s| !state.cast_spells.contains(s) && s.cost <= state.player.mana).map(|s| **s).collect()
}

fn find_best_mana(initial_state: State, hard_mode: bool) -> i32 {
    let mut q: Vec<State> = vec![initial_state];

    let mut best = 99_999;
    while !q.is_empty() {
        let mut current = q.pop().unwrap();
        if hard_mode {
            current.player.hp -= 1;
            if check_match_over(&current, &mut best) {
                continue;
            }
        }
        tick_spells(&mut current);

        if check_match_over(&current, &mut best) {
            continue;
        }
        
        for next_spell in get_next_spells(current.clone()).iter() {
            let mut next = current.clone();
            next.cast_spells.push(*next_spell);
            next.player.mana -= next_spell.cost;
            next.mana_spent += next_spell.cost;
            if next.mana_spent >= best {
                continue;
            }

            //Boss Turn
            tick_spells(&mut next);
            if check_match_over(&next, &mut best) {
                continue;
            }

            next.player.hp -= match next.boss.atk - next.player.armor {
                dmg if dmg < 1 => 1,
                dmg => dmg
            };
            if check_match_over(&next, &mut best) {
                continue;
            }

            q.push(next);
        }
    }

    best
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_one() {
        let player = Player{hp: 10, mana: 250, armor: 0};
        let boss = Boss{hp: 13, atk: 8};
        let initial_state = State{player: player, boss: boss, cast_spells: Vec::new(), mana_spent: 0};

        assert_eq!(226, find_best_mana(initial_state, false));
    }

    #[test]
    fn example_two() {
        let player = Player{hp: 10, mana: 250, armor: 0};
        let boss = Boss{hp: 14, atk: 8};
        let initial_state = State{player: player, boss: boss, cast_spells: Vec::new(), mana_spent: 0};

        assert_eq!(641, find_best_mana(initial_state, false));
    }

    //MagicMissle, Drain, Shield, Poison, Recharge
    //    0          1       2       3        4

    #[test]
    fn can_cast_tests() {
        let mut state = State{ player: Player{hp: 10, mana: 1000, armor: 0}, boss: Boss{hp: 14, atk: 8}, cast_spells: Vec::new(), mana_spent: 0 };

        let castable = get_next_spells(state.clone());
        dbg!(&castable);
        assert!(castable.contains(ALL_SPELLS[2]));

        state.cast_spells.push(*ALL_SPELLS[2]); //Shield
        tick_spells(&mut state); //5
        tick_spells(&mut state); //4

        let castable = get_next_spells(state.clone());
        dbg!(&castable);
        assert!(!castable.contains(ALL_SPELLS[2]));
        tick_spells(&mut state); //3
        tick_spells(&mut state); //2

        let castable = get_next_spells(state.clone());
        assert!(!castable.contains(ALL_SPELLS[2]));
        tick_spells(&mut state); //1
        tick_spells(&mut state); //0

        let castable = get_next_spells(state.clone());
        assert!(castable.contains(ALL_SPELLS[2]));
    }

    #[test]
    fn tick_spells_tests() {
        let mut state = State{ player: Player{hp: 10, mana: 0, armor: 0}, boss: Boss{hp: 14, atk: 8}, cast_spells: Vec::new(), mana_spent: 0 };

        // Turn 1
        state.cast_spells.push(*ALL_SPELLS[4]); //Recharge
        tick_spells(&mut state);
        assert_eq!(101, state.player.mana);

        tick_spells(&mut state);
        assert_eq!(202, state.player.mana);

        // Turn 2
        state.cast_spells.push(*ALL_SPELLS[2]); //Shield
        tick_spells(&mut state);
        assert_eq!(303, state.player.mana);
        assert_eq!(7, state.player.armor);

        tick_spells(&mut state);
        assert_eq!(404, state.player.mana);
        assert_eq!(7, state.player.armor);

        // Turn 3
        state.cast_spells.push(*ALL_SPELLS[1]); //Drain
        tick_spells(&mut state);
        assert_eq!(505, state.player.mana);
        assert_eq!(7, state.player.armor);
        assert_eq!(12, state.player.hp);

        tick_spells(&mut state);
        assert_eq!(505, state.player.mana);
        assert_eq!(7, state.player.armor);
        assert_eq!(12, state.player.hp);

        // Turn 4
        state.cast_spells.push(*ALL_SPELLS[3]); //Poison
        tick_spells(&mut state);
        assert_eq!(7, state.player.armor);

        tick_spells(&mut state);
        assert_eq!(7, state.player.armor);

        // Turn 5
        state.cast_spells.push(*ALL_SPELLS[0]); //Magic Missle
        tick_spells(&mut state);
        assert_eq!(0, state.player.armor);
    }
}

File: Cpp/AdventOfCode/src/2015/d22_BossBattle2.cpp
#pragma once

#include "2015/d22_BossBattle2.h"

SOLUTION(2015, 22) {
    struct Player { s32 Hp{ 0 }; s32 Mana{ 0 }; s32 Armor{ 0 }; };
    struct Boss { s32 Hp{ 0 }; s32 Atk{ 0 }; };
    constexpr Boss ParseBoss(const auto& lines) {
        Boss result;
        auto hp = Constexpr::Split(lines[0], ": ")[1];
        auto dmg = Constexpr::Split(lines[1], ": ")[1];

        Constexpr::ParseNumber(hp, result.Hp);
        Constexpr::ParseNumber(dmg, result.Atk);
        return result;
    }

    enum struct Spell {
        BossTurn, //hack?
        MagicMissle,
        Drain,
        Shield,
        Poison,
        Recharge
    };

    constexpr size_t GetDuration(Spell spell) {
        switch (spell) {
        case Spell::Shield:
        case Spell::Poison: return 6;
        case Spell::Recharge: return 5;
        default: return 1;
        }
    }
    constexpr s32 GetManaCost(Spell spell) {
        switch (spell) {
        case Spell::MagicMissle: return 53;
        case Spell::Drain: return 73;
        case Spell::Shield: return 113;
        case Spell::Poison: return 173;
        case Spell::Recharge: return 229;
        default: return 0;
        }
    }
    struct State {
        Player Player;
        Boss Boss;
        std::vector<Spell> CastSpells;
    };

    constexpr bool CanCast(const State& state, Spell spell) {
        if (GetManaCost(spell) > state.Player.Mana) return false;
        if (GetDuration(spell) == 1) return true;

        auto offset = std::min(state.CastSpells.size(), GetDuration(spell) - 1);
        return std::find(state.CastSpells.cend() - offset, state.CastSpells.cend(), spell) == state.CastSpells.cend();
    }

    constexpr bool TestCanCast() {
        State state{ {10, 1000, 0}, {14, 8}, {} };
        if (!CanCast(state, Spell::Shield)) return false;
        state.CastSpells.push_back(Spell::BossTurn);
        state.CastSpells.push_back(Spell::Shield); //5

        if (CanCast(state, Spell::Shield)) return false;
        state.CastSpells.push_back(Spell::BossTurn); //4
        state.CastSpells.push_back(Spell::MagicMissle); //3

        if (CanCast(state, Spell::Shield)) return false;
        state.CastSpells.push_back(Spell::BossTurn); //2
        state.CastSpells.push_back(Spell::MagicMissle); //1

        if (CanCast(state, Spell::Shield)) return false;
        state.CastSpells.push_back(Spell::BossTurn); //0

        if (!CanCast(state, Spell::Shield)) return false;

        return true;
    }

    constexpr s32 CalculateMana(const State& state) {
        s32 result = 0;
        for (const auto& spell : state.CastSpells) {
            result += GetManaCost(spell);
        }
        return result;
    }

    constexpr void TickSpell(Spell spell, State& state) {
        switch (spell) {
        case Spell::MagicMissle: state.Boss.Hp -= 4; return;
        case Spell::Drain: state.Player.Hp += 2; state.Boss.Hp -= 2; return;
        case Spell::Shield: state.Player.Armor = 7; return;
        case Spell::Poison: state.Boss.Hp -= 3; return;
        case Spell::Recharge: state.Player.Mana += 101; return;
        }
    }
    constexpr void TickSpells(State& state) {
        state.Player.Armor = 0;
        for (size_t i = 0; i < state.CastSpells.size(); i++) {
            auto delta = state.CastSpells.size() - i;
            auto spell = state.CastSpells[i];
            auto duration = GetDuration(spell);
            if (duration < delta) continue;
            TickSpell(spell, state);
        }
    }

    constexpr bool TestTickSpells() {
        State state{ {10, 0, 0}, {14, 8}, {} };
        
        // Turn 1
        state.CastSpells.push_back(Spell::Recharge);
        TickSpells(state);
        if (state.Player.Mana != 101) return false;
        
        state.CastSpells.push_back(Spell::BossTurn);
        TickSpells(state);
        if (state.Player.Mana != 202) return false;

        // Turn 2
        state.CastSpells.push_back(Spell::Shield);
        TickSpells(state);
        if (state.Player.Mana != 303) return false;
        if (state.Player.Armor != 7) return false;

        state.CastSpells.push_back(Spell::BossTurn);
        TickSpells(state);
        if (state.Player.Mana != 404) return false;
        if (state.Player.Armor != 7) return false;

        // Turn 3
        state.CastSpells.push_back(Spell::Drain);
        TickSpells(state);
        if (state.Player.Mana != 505) return false;
        if (state.Player.Armor != 7) return false;
        if (state.Player.Hp != 12) return false;

        state.CastSpells.push_back(Spell::BossTurn);
        TickSpells(state);
        if (state.Player.Mana != 505) return false;
        if (state.Player.Armor != 7) return false;
        if (state.Player.Hp != 12) return false; //don't invoke drain twice

        // Turn 4
        state.CastSpells.push_back(Spell::Poison);
        TickSpells(state);
        if (state.Player.Armor != 7) return false;
        
        state.CastSpells.push_back(Spell::BossTurn);
        TickSpells(state);
        if (state.Player.Armor != 7) return false;

        // Turn 5
        state.CastSpells.push_back(Spell::MagicMissle);
        TickSpells(state);
        if (state.Player.Armor != 0) return false;
        return true;
    }

    constexpr s32 FindBestMana(const State& initialState, bool hardMode) {
        std::vector<State> q{ initialState };
        std::vector<State> nextQ;
        std::vector<Spell> allSpells = { Spell::MagicMissle, Spell::Drain, Spell::Shield, Spell::Poison, Spell::Recharge };
        s32 bestMana = 99999;

        while (!q.empty()) {
            for (const auto& state : q) {
                auto currentManaCost = CalculateMana(state);
                if (currentManaCost > bestMana) continue;

                State current = state;
                if (hardMode) {
                    current.Player.Hp--;
                    if (current.Player.Hp == 0) continue;
                }
                
                current.CastSpells.push_back(Spell::BossTurn);
                TickSpells(current);
                if (current.Boss.Hp <= 0) {
                    bestMana = std::min(bestMana, currentManaCost);
                    continue;
                }

                for (auto spell : allSpells) {
                    if (!CanCast(current, spell)) continue;
                    State next = current;
                    next.CastSpells.push_back(spell);
                    next.Player.Mana -= GetManaCost(spell);

                    //Boss Turn
                    TickSpells(next);
                    if (next.Boss.Hp <= 0) {
                        bestMana = std::min(bestMana, currentManaCost + GetManaCost(spell));
                        continue;
                    }
                    next.Player.Hp -= std::max(1, next.Boss.Atk - next.Player.Armor);
                    if (next.Player.Hp <= 0) continue;

                    nextQ.push_back(next);
                }
            }
            q.clear();
            std::copy(nextQ.cbegin(), nextQ.cend(), std::back_inserter(q));
            nextQ.clear();
        }

        return bestMana;
    }
    PART_ONE() {
        auto boss = ParseBoss(lines);
        Player player = { 50, 500, 0 };
        State initialState{ player, boss, {} };
        return Constexpr::ToString(FindBestMana(initialState, false));
    }
    PART_TWO() {
        auto boss = ParseBoss(lines);
        Player player = { 50, 500, 0 };
        State initialState{ player, boss, {} };
        return Constexpr::ToString(FindBestMana(initialState, true));
    }
    TESTS() {
        if (!TestCanCast()) return false;
        if(!TestTickSpells()) return false;

        Player player = { 10, 250, 0 };
        Boss boss = { 13, 8 };
        State initialState{ player, boss, {} };
        if (FindBestMana(initialState, false) != 226) return false;

        player = { 10, 250, 0 };
        boss = { 14, 8 };
        initialState = { player, boss, {} };
        if (FindBestMana(initialState, false) != 641) return false;

        return true;
    }
}